### PR TITLE
nar/serialize.go: fix reader

### PR DESCRIPTION
**Pull request by flokli**
There's no guarantees on how many bytes the Read() method of an io.Reader will write.
We can't assume it to block until all bytes requested are there - it can
return before (especially if it's still receiving on the other side).

This adds a "hesitantReader", which behaves like this, a failing unit test, then switches serialize.go over to use io.ReadFull (which behaves like we want it to).

After all that is done, we switch over to encoding/binary instead of bitshifting manually to produce the (u)int64.

### CODE REVIEW

**Comment by flokli** [comment]
Note the other `read*` methods suffer the same problem - I'll address that in a followup.

File: src/nar/serialize.go [hunk 1]
```diff
@@ -1,10 +1,15 @@
 package nar
 
 import (
+	"encoding/binary"
 	"fmt"
 	"io"
 )
 
+const (
+	MaxInt64 = 1<<(64-1) - 1
```
Comment (zimbatm) [inline src/nar/serialize.go:10]: `math.MaxInt64` also has that value. But maybe you didn't want to import the whole module just for the constant.

File: src/nar/serialize.go [hunk 2]
```diff
@@ -61,32 +66,19 @@ func readPadding(r io.Reader, l int64) error {
 const maxInt64 = 1<<63 - 1
```
Comment (zimbatm) [inline src/nar/serialize.go:66]: Is that a duplicate of the above? With a different value!

**Review by zimbatm — COMMENTED** [review]
Good find!

File: src/nar/serialize.go (hunk 2 above)
Comment (flokli) [reply src/nar/serialize.go:66]: I think 64-1 also is 63, but yeah, I missed this constant being present :-)

Will update.

File: src/nar/serialize.go (hunk 1 above)
Comment (flokli) [reply src/nar/serialize.go:10]: I think the go compiler is pretty good in dead code elimination. I'll use `math.MaxInt64`, thanks!

**Pull request merged**